File: faiss/textIndexer.js
const tf = require("@tensorflow/tfjs-node");
const use = require("@tensorflow-models/universal-sentence-encoder");
const fs = require("fs").promises;
const path = require("path");

class TextIndexer {
  constructor() {
    this.stateFile = "index_state.json";
    this.vectorFile = "vectors.json";
    this.model = null;
    this.vectors = [];
    this.state = {
      files: {}, // Maps file paths to last modified times
      sentences: [], // Original sentences
      fileToIndices: {}, // Maps file paths to vector indices
    };
  }

  async initialize() {
    // Load Universal Sentence Encoder model
    this.model = await use.load();

    // Load existing state and vectors if they exist
    try {
      const stateExists = await fs
        .access(this.stateFile)
        .then(() => true)
        .catch(() => false);

      if (stateExists) {
        const stateData = await fs.readFile(this.stateFile, "utf-8");
        this.state = JSON.parse(stateData);

        const vectorData = await fs.readFile(this.vectorFile, "utf-8");
        this.vectors = JSON.parse(vectorData);
      }
    } catch (error) {
      console.error("Error loading existing state:", error);
      // Continue with empty state if loading fails
    }
  }

  async saveState() {
    await fs.writeFile(this.stateFile, JSON.stringify(this.state, null, 2));
    await fs.writeFile(this.vectorFile, JSON.stringify(this.vectors));
  }

  async getSentencesFromFile(filePath) {
    const content = await fs.readFile(filePath, "utf-8");
    // Split text into sentences (simple implementation)
    return content
      .split(/[.!?]+/)
      .map((s) => s.trim())
      .filter((s) => s.length > 0);
  }

  needsUpdate(filePath) {
    if (!this.state.files[filePath]) {
      return true;
    }

    const stats = fs.statSync(filePath);
    return stats.mtime.getTime() > this.state.files[filePath];
  }

  async updateIndex(directory) {
    const files = await fs.readdir(directory);

    for (const file of files) {
      if (!file.endsWith(".txt")) continue;

      const filePath = path.join(directory, file);
      if (!this.needsUpdate(filePath)) continue;

      // Get sentences from file
      const sentences = await this.getSentencesFromFile(filePath);
      if (sentences.length === 0) continue;

      // Generate embeddings
      const embeddings = await this.model.embed(sentences);
      const vectors = await embeddings.array();

      // Remove old vectors if file was previously indexed
      if (this.state.fileToIndices[filePath]) {
        const [startIdx, count] = this.state.fileToIndices[filePath];
        this.vectors.splice(startIdx, count);
        this.state.sentences.splice(startIdx, count);

        // Update indices for files that come after
        for (const [path, [idx, cnt]] of Object.entries(
          this.state.fileToIndices
        )) {
          if (idx > startIdx) {
            this.state.fileToIndices[path] = [
              idx - count + vectors.length,
              cnt,
            ];
          }
        }
      }

      // Add new vectors and update state
      const startIdx = this.vectors.length;
      this.vectors.push(...vectors);
      this.state.sentences.push(...sentences);
      this.state.fileToIndices[filePath] = [startIdx, vectors.length];
      this.state.files[filePath] = Date.now();
    }

    await this.saveState();
  }

  async search(query, k = 5) {
    if (!this.model || this.vectors.length === 0) {
      return [];
    }

    // Generate query embedding
    const queryEmbedding = await this.model.embed([query]);
    const queryVector = await queryEmbedding.array();

    // Calculate cosine similarity with all vectors
    const similarities = this.vectors.map((vector) => {
      const dotProduct = tf.tensor1d(queryVector[0]).dot(tf.tensor1d(vector));
      const queryNorm = tf.tensor1d(queryVector[0]).norm();
      const vectorNorm = tf.tensor1d(vector).norm();
      const similarity = dotProduct.div(queryNorm.mul(vectorNorm));
      return similarity.arraySync();
    });

    // Get top k results
    const results = similarities
      .map((score, index) => ({ score, text: this.state.sentences[index] }))
      .sort((a, b) => b.score - a.score)
      .slice(0, k);

    return results;
  }
}

// Example usage
async function main() {
  const indexer = new TextIndexer();
  await indexer.initialize();

  // Update index with all txt files in data directory
  console.log("Updating index...");
  await indexer.updateIndex("data");

  // Example search
  const query = "machine learning and artificial intelligence";
  console.log(`\nSearching for: ${query}`);
  const results = await indexer.search(query, 3);

  console.log("\nResults:");
  for (const { score, text } of results) {
    console.log(`\nScore: ${score.toFixed(4)}`);
    console.log(`Text: ${text}`);
  }
}

if (require.main === module) {
  main().catch(console.error);
}

module.exports = TextIndexer;

File: faiss/text_indexer.py
import os
import json
import faiss
import numpy as np
from sentence_transformers import SentenceTransformer
from typing import List, Dict, Tuple
from datetime import datetime

class TextIndexer:
    def __init__(self, model_name: str = 'all-MiniLM-L6-v2'):
        """Initialize the text indexer with a sentence transformer model."""
        self.model = SentenceTransformer(model_name)
        self.dimension = 384  # Output dimension of the model
        self.index = faiss.IndexFlatL2(self.dimension)
        self.state_file = 'last_index.json'
        self.index_file = 'text_vectors.faiss'
        self.state = self._load_state()
        self._load_or_create_index()

    def _load_state(self) -> Dict:
        """Load or create the indexing state."""
        if os.path.exists(self.state_file):
            with open(self.state_file, 'r') as f:
                return json.load(f)
        return {
            'files': {},  # Maps file paths to last modified times
            'total_vectors': 0,
            'file_to_indices': {}  # Maps file paths to their vector indices
        }

    def _save_state(self):
        """Save the current indexing state."""
        with open(self.state_file, 'w') as f:
            json.dump(self.state, f, indent=2)

    def _load_or_create_index(self):
        """Load existing index or create a new one."""
        if os.path.exists(self.index_file):
            self.index = faiss.read_index(self.index_file)
        else:
            self.index = faiss.IndexFlatL2(self.dimension)

    def _save_index(self):
        """Save the FAISS index to disk."""
        faiss.write_index(self.index, self.index_file)

    def _get_file_sentences(self, file_path: str) -> List[str]:
        """Split file content into sentences."""
        with open(file_path, 'r') as f:
            content = f.read()
        # Simple sentence splitting on periods followed by whitespace
        sentences = [s.strip() for s in content.split('.') if s.strip()]
        return sentences

    def _needs_update(self, file_path: str) -> bool:
        """Check if a file needs to be indexed or updated."""
        if not os.path.exists(file_path):
            return False
        
        mtime = os.path.getmtime(file_path)
        return (file_path not in self.state['files'] or 
                mtime > self.state['files'][file_path])

    def update_index(self, directory: str):
        """Update the index with new or modified files."""
        for root, _, files in os.walk(directory):
            for file in files:
                if not file.endswith('.txt'):
                    continue
                
                file_path = os.path.join(root, file)
                if not self._needs_update(file_path):
                    continue

                # Remove old vectors for this file if it exists
                if file_path in self.state['file_to_indices']:
                    old_start, old_count = self.state['file_to_indices'][file_path]
                    # Note: In a production system, we'd need to handle deletion properly
                    # For this example, we'll just overwrite the old vectors

                sentences = self._get_file_sentences(file_path)
                if not sentences:
                    continue

                # Convert sentences to embeddings
                embeddings = self.model.encode(sentences, convert_to_numpy=True)
                embeddings = np.ascontiguousarray(embeddings.astype('float32'))

                # Add new vectors
                start_idx = self.state['total_vectors']
                self.index.add(embeddings)

                # Update state
                self.state['files'][file_path] = os.path.getmtime(file_path)
                self.state['file_to_indices'][file_path] = (start_idx, len(sentences))
                self.state['total_vectors'] += len(sentences)

        # Save updated state and index
        self._save_state()
        self._save_index()

    def search(self, query: str, k: int = 5) -> List[Tuple[str, float]]:
        """Search for similar content."""
        # Convert query to embedding
        query_vector = self.model.encode([query], convert_to_numpy=True)
        query_vector = np.ascontiguousarray(query_vector.astype('float32'))

        # Search the index
        distances, indices = self.index.search(query_vector, k)
        
        # Find corresponding files and sentences
        results = []
        for idx, dist in zip(indices[0], distances[0]):
            # Find which file this vector belongs to
            for file_path, (start_idx, count) in self.state['file_to_indices'].items():
                if start_idx <= idx < start_idx + count:
                    # Read the specific sentence
                    sentences = self._get_file_sentences(file_path)
                    sentence_idx = idx - start_idx
                    if sentence_idx < len(sentences):
                        results.append((sentences[sentence_idx], dist))
                    break

        return results

if __name__ == "__main__":
    # Example usage
    indexer = TextIndexer()
    
    # Update index with all txt files in data directory
    print("Updating index...")
    indexer.update_index('data')
    
    # Example search
    query = "machine learning and artificial intelligence"
    print(f"\nSearching for: {query}")
    results = indexer.search(query, k=3)
    
    print("\nResults:")
    for text, score in results:
        print(f"Score: {score:.4f}")
        print(f"Text: {text}")
        print()